perf_bench_test.go

```diff
@@ -118,83 +130,122 @@ func TestBenchRequestPerSecond(t *testing.T) {
 	var (
-		vmID, requestID             int
-		invokExecTime, realRequests int64
-		vmGroup                     sync.WaitGroup
-		isProfile                   = false
-		timeInterval                = time.Duration(time.Second.Nanoseconds() / int64(targetRPS))
-		injectDuration              = *warmUpTime + *profileTime + *coldDownTime
-		totalRequests               = injectDuration * float64(targetRPS)
-		remainingRequests           = totalRequests
-		ticker                      = time.NewTicker(timeInterval)
-		profiler                    = profile.NewProfiler(injectDuration, *profilerInterval, vmNum, *profilerLevel, *profilerNodes, "profile")
+		pmuMetric      *metrics.Metric
+		vmGroup        sync.WaitGroup
+		isProfile      = false
+		cores          = runtime.NumCPU()
+		threshold      = 5 * getUnloadServiceTime()
+		injectDuration = *warmUpTime + *profileTime + *coolDownTime
 	)
 
 	const (
 		averageExecutionTime  = "average-execution-time"
 		realRequestsPerSecond = "real-requests-per-second"
 	)
 
-	tStart := time.Now()
-	err := profiler.Run()
-	require.NoError(t, err, "Run profiler returned error")
-	go profileControl(vmNum, &isProfile, profiler)
+	for i := 1.; i < 9; i++ {
+		var (
+			vmID, requestID             int
+			invokExecTime, realRequests int64
+			serveMetric                 = metrics.NewMetric()
+			rps                         = 0.1 * i * float64(targetRPS)
+			timeInterval                = time.Duration(time.Second.Nanoseconds() / int64(rps))
+			ticker                      = time.NewTicker(timeInterval)
+			totalRequests               = injectDuration * rps
+			remainingRequests           = totalRequests
+			profiler                    = profile.NewProfiler(injectDuration, *profilerInterval, vmNum, *profilerLevel, *profilerNodes, "profile", *isBindSocket)
+		)
 
-	for remainingRequests > 0 {
-		if tickerT := <-ticker.C; !tickerT.IsZero() {
-			vmGroup.Add(1)
-			remainingRequests--
+		log.Debugf("RPS: %f", rps)
+
+		tStart := time.Now()
+		latencies := make(chan float64, 2)
+		go latencyMeasurement(t, 1, images[1%len(images)], &isProfile, latencies)
+
+		err := profiler.Run()
+		require.NoError(t, err, "Run profiler returned error")
+		go profileControl(vmNum, &isProfile, profiler)
+
+		for remainingRequests > 0 {
+			if tickerT := <-ticker.C; !tickerT.IsZero() {
+				vmGroup.Add(1)
+				remainingRequests--
+
+				imageName := images[vmID%len(images)]
 
-			imageName := images[vmID%len(images)]
+				go serveVM(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
+				requestID++
 
-			go serveVM(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
-			requestID++
+				vmID = (vmID + 1) % vmNum
+			}
+		}
+		ticker.Stop()
+		vmGroup.Wait()
+		log.Debugf("All VM returned in %d Milliseconds", time.Since(tStart).Milliseconds())
+		meanLatency, percentileLatency := <-latencies, <-latencies
+		log.Debug(meanLatency, percentileLatency)
```

Comment:
please add a message explaining what is printed in the logs

Reply:
Done